azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java

```diff
@@ -696,8 +705,13 @@ private Status runJob() {
         logError("Job run failed, but will treat it like success.");
         logError(e.getMessage() + " cause: " + e.getCause(), e);
       } else {
-        finalStatus = changeStatus(Status.FAILED);
-        logError("Job run failed!", e);
+        if (node.getStatus() == Status.KILLED) {
```

Comment:
I think it's the best bet to keep it.

I'm only improving the "exit status of `runJob()` method", but it's also possible that the job is killed before `runJob()` is called at all. And I don't think status setting is synchronized well enough at the moment, I believe if the timing of `kill()` overlaps with `runJob()` in a certain way, it could still cause the flapping of `KILLED->FAILED->KILLED`.

Those possible mis-sync issues are really rare corner cases though. My change here was enough to keep the test passing without random failures. If someone wants to take it further, that could be done in another PR.